Review

aleoreh opened the pull request.

aleoreh committed:
Реализует поведение открытия и закрытия формы

 - параметров

aleoreh committed:
Исправляет размещение слушателя ArticleParamsForm

aleoreh committed:
Пытается исправить поведение lint-staged

aleoreh committed:
Добавляет функцию Debug.log

aleoreh committed:
Увеличивает строгость правила eslint

- теперь недопустимо использовать debug

aleoreh committed:
Сокращает код debug.ts

aleoreh committed:
Реализует закрытие формы параметров

- ArticleParamsForm теперь закрывается
с помощью useOutsideClick hook

aleoreh committed:
Добавляет состояние "articleStyle" в компонент App

aleoreh committed:
Добавляет в App состояние для введенных стилей

aleoreh committed:
Добавляет компонент InputArticleParams

aleoreh committed:
Выводит содержимое InputArticleParams в форму

aleoreh committed:
Выносит тип ArticleStyle в отдельный модуль

aleoreh committed:
Реализует установку состояния InputArticleStyle

aleoreh committed:
Реорганизует тип хука useOutsideClick

aleoreh committed:
Реорганизует тип InputArticleParams

aleoreh committed:
+ заголовок формы

aleoreh committed:
Останавливает click из ArticleParamsForm

aleoreh committed:
Добавляет на форму выбор семейства шрифтов

aleoreh committed:
Добавляет на форму выбор размера шрифта

aleoreh committed:
Добавляет на форму выбор цвета текста

aleoreh committed:
Добавляет на форму выбор цвета фона

aleoreh committed:
Добавляет на форму выбор ширины контента

aleoreh committed:
Реализует возможность сброса введенных значений

aleoreh committed:
Исправляет ошибку радиокнопок

- в некоторых случаях невозможно было выбрать опцию

aleoreh committed:
Добавляет отступы полям ввода

aleoreh committed:
Организует импорты

aleoreh committed:
Добавляет некоторые комментарии к коду

aleoreh committed:
Организует код

aleoreh committed:
Устанавливает поведение при нажатии на сброс

aleoreh committed:
Добавляет описание компонента InputArticleParams

aleoreh committed:
Переименовывает функции для соблюдения стиля

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.module.scss:
```diff
@@ -30,3 +30,7 @@
 	margin-top: auto;
 }
+
+.form_spacing {
```
**Можно лучше:**
В этой проектной работе нет необходимости писать дополнительный css-класс. Свойство `gap` просто добавляем в класс `.form`.

aleoreh replied:
Старался не затрагивать изначальный код. Исправлено

olgalatkina commented on src/index.tsx:
```diff
@@ -1,31 +1,63 @@
+
 const App = () => {
+	// стиль, который применяется к статье
+	const [articleStyle, setArticeStyle] = useState<ArticleStyle>(defaultStyle);
```
**Надо исправить:** 
Настройки статьи по умолчанию описаны в объекте `defaultArticleState`  в файле `constants/articleProps.ts` стартеркита. В проекте нужно использовать значения из этого объекта как настройки по умолчанию, как для самой статьи, так и как начальные значения полей форм.

olgalatkina commented on src/types.ts:
```diff
@@ -0,0 +1,7 @@
+export type ArticleStyle = {
```
**Надо исправить:** 
В проекте как настройки по умолчанию нужно использовать значения из объекта `defaultArticleState`. Этот тип стоит удалить как неиспользуемый.

olgalatkina commented on src/index.tsx:
```diff
@@ -1,31 +1,63 @@
+	'--bg-color': defaultArticleState.backgroundColor.value,
+};
+
 const App = () => {
```
Обычно корневой `index.tsx` держат максимально компактным. Стоит вынести компонент `App` в отдельный файл в папке `components/app/App.tsx`.

olgalatkina commented on src/index.tsx:
```diff
@@ -1,31 +1,63 @@
+	const [articleStyle, setArticeStyle] = useState<ArticleStyle>(defaultStyle);
+
+	// стиль, который вводится пользователем
+	const [inputArticleStyle, setInputArticleStyle] =
```
**Надо исправить:** 
Каждый компонент сам должен отвечать за хранение и обработку своего состояния. Использовать подъем состояния в родительский компонент не имеет смысла для нашего приложения. Состояние формы нужно описать в самом компонент `ArticleParamsForm`, а не передавать его из `App`, т.к. это состояние описывает именно `ArticleParamsForm`. В состоянии `App` должны хранится только данные о текущем отображении статьи.

aleoreh replied:
Исправлено вместе с избавлением от InputArticleParams

olgalatkina commented on src/index.tsx:
```diff
@@ -1,31 +1,63 @@
+		useState<ArticleStyle>(articleStyle);
+
+	// видимость формы параметорв
+	const [articleParamsFormOpened, setArticleParamsFormOpened] =
```
**Надо исправить:**
1. Обычно переменные, которые сохраняют булево значение начинаются с глагола `is` или `has`, например, `isOpen`
2. Это состояние тоже нужно перенести в компонент `ArticleParamsForm` вместе обработчиками отправки и сброса формы.

olgalatkina commented on src/index.tsx:
```diff
@@ -1,31 +1,63 @@
 const domNode = document.getElementById('root') as HTMLDivElement;
 const root = createRoot(domNode);
 
+const defaultStyle: ArticleStyle = {
```
**Надо исправить:**
В проекте как настройки по умолчанию нужно использовать значения из объекта `defaultArticleState`. Этот объект стоит удалить как неиспользуемый.

aleoreh replied:
Исправлено вместе с "Переключает состояние по умолчанию"

olgalatkina commented on src/index.tsx:
```diff
@@ -1,31 +1,63 @@
+				submit={submitArticleParamsForm}
+				reset={resetArticleParamsForm}>
+				{/* поля ввода параметров передаются как "дети" формы */}
+				<InputArticleParams
```
**Надо исправить:**
Поля формы это часть компонента `ArticleParamsForm`, отвечающего за настройки отображения статьи и не предусмотрено, что в форму будет передаваться разный набор полей ввода. 
В этой проектной работе всё наполнение формы прописываем в компоненте `ArticleParamsForm`. Просто собираем её из готовых компонентов. Компонент `InputArticleParams` не понадобится.

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,14 +1,67 @@
+				}}>
+				<form
+					className={clsx(styles.form, styles.form_spacing)}
+					onSubmit={submitForm}
```
**Отлично:**
Отлично, что обработчик сохранения данных вызывается при отправке формы.

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,14 +1,67 @@
+		props.setOpen(!props.open);
+	};
+
+	const submitForm = (evt: React.FormEvent) => {
```
**Можно лучше:**
`React.FormEvent<HTMLFormElement>`

olgalatkina commented on src/components/arrow-button/ArrowButton.tsx:
```diff
@@ -1,19 +1,42 @@
+	onClick: OnClick;
 };
+
+export const ArrowButton = forwardRef<HTMLDivElement, ArrowButtonProps>(
```
**Нужно исправить:**
Использование `forwardRef` избыточно. Достаточно передать флаг открытости/закрытости сайдбара и функцию меняющую этот флаг.

olgalatkina commented on src/components/arrow-button/ArrowButton.tsx:
```diff
@@ -1,19 +1,42 @@
+				role='button'
+				aria-label='Открыть/Закрыть форму параметров статьи'
+				tabIndex={0}
+				className={clsx(
```
**Отлично:**
Здорово, что используете библиотеку `clsx`.

olgalatkina commented on src/components/arrow-button/ArrowButton.tsx:
```diff
@@ -1,19 +1,42 @@
 };
+
+export const ArrowButton = forwardRef<HTMLDivElement, ArrowButtonProps>(
+	(props: ArrowButtonProps, ref) => {
```
**Можно лучше:** 
Можно использовать деструктуризацию пропсов сразу в параметрах компонента, это распространенная практика.

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,14 +1,67 @@
+		onClick: () => props.setOpen(false),
+	});
+
+	const toggleOpen = () => {
```
**Можно лучше:** 
Нет необходимости создавать дополнительные функции обработчики если они сами по себе ничего не делают, можно использовать просто функции сеттеры возвращаемые `useState`.

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,14 +1,67 @@
+
+export const ArticleParamsForm = (props: ArticleParamsFormProps) => {
+	// ссылка на элемент контейнера для формы
+	const containerRef = useRef<HTMLElement>(null);
```
**Нужно исправить:**
Реф инициализируется значением `null`, поэтому типизация будет выглядеть так:
```
const containerRef = useRef<HTMLElement | null>(null);
```

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,14 +1,67 @@
+	children: JSX.Element;
+};
+
+export const ArticleParamsForm = (props: ArticleParamsFormProps) => {
```
**Можно лучше:**
Можно использовать деструктуризацию пропсов сразу в параметрах компонента, это распространенная практика.

olgalatkina commented on src/hooks/useOutsideClick.ts:
```diff
@@ -0,0 +1,30 @@
+			}
+		};
+
+		window.addEventListener('click', handleClick);
```
**Можно лучше:**
Используйте событие `mousedown` и тогда не придётся использовать `evt.stopPropagation()` у компонента `ArrowButton`. 
Таким же образом стоит "починить" хук `useOutsideClickClose` в компоненте `Select`.

**Отлично:**
Здорово, что вы вынесли логику закрытия сайдбара по клику вне сайдбара в кастомный хук.
Можете попробовать другое готовое решение, в нём так же реализовано закрытие по `Escape`:
```
import { useEffect } from "react";

type TUseClose = {
  isOpen: boolean;
  onClose: () => void;
  rootRef: React.RefObject<HTMLElement>;
};

// Кастомные хуки всегда должны начинаться с глагола `use`, чтобы реакт понял, что это хук. 
// Он следит за их вызовами
export function useClose({isOpen, onClose, rootRef}: TUseClose) {
  useEffect(() => {
    if (!isOpen) return; // останавливаем действие эффекта, если закрыто
   
    function handleClickOutside(event: MouseEvent) {
      const { target } = event;
      
      const isOutsideClick =
        target instanceof Node && // проверяем, что это `DOM`-элемент
        rootRef.current &&
        !rootRef.current.contains(target); // проверяем, что кликнули на элемент, который находится не внутри блока
      if (isOutsideClick) {
        onClose();
      }
    }

    const handleEscape = (event: KeyboardEvent) => {
      if (event.key === "Escape") {
        onClose();
      }
    };

    document.addEventListener("keydown", handleEscape);
    document.addEventListener("mousedown", handleClickOutside);

    //  Обязательно удаляем обработчики в `clean-up`- функции
    return () => {
      document.removeEventListener("keydown", handleEscape);
      document.removeEventListener("mousedown", handleClickOutside);
    };
    
    // Обязательно следим за `isOpen`, чтобы срабатывало только при открытии, 
    // а не при любой перерисовке компонента
  }, [isOpen, onClose, rootRef]);
}
```

Этот код помещают в отдельный файл `useClose.ts` в папке `hooks` (она у вас уже создана) и импортируют хук туда, где нужно.
Теперь его можно использовать в любом компоненте (попапе, меню и тд), которому нужно установить эти обработчики.

Пример того, как это использовать:
```
    // Создаем `ref` для формы и указываем его на внешнем блоке с формой
    const formRef = useRef<HTMLElement | null>(null);
    
    useClose({
        isOpen,
        onClose: () => setIsOpen(false),
        rootRef: formRef,
    }); 
```

aleoreh replied:
Реализовал хук useModalClose с отработкой mousedown и keydown

olgalatkina commented on src/hooks/useOutsideClick.ts:
```diff
@@ -0,0 +1,30 @@
+ * Выполняет onClick при клике за пределами переданных ссылок rootRefs
+ */
+export const useOutsideClick: UseOutsideClick = ({ rootRefs, onClick }) => {
+	useEffect(() => {
```
**Надо исправить:** 
Обработчик не должен висеть всегда на документе, его нужно вешать в момент открытия и удалять в момент закрытия. Нет необходио отслеживать событие, если сайдбар закрыт. Поэтому необходимо указать массив зависимостей хука `useEffect` и добавить внутри проверку, что сайдбар открыт. 
Чтобы это реализовать, вам понадобится передать в хук флаг `isOpen`.

olgalatkina commented on src/components/input-article-params/InputArticleParams.tsx:
```diff
@@ -0,0 +1,111 @@
+				title='Шрифт'
+				options={fontFamilyOptions}
+				selected={defaultFont}
+				onChange={onFontSelected}
```
**Можно лучше:** 
Если для каждого поля ввода формы создается свой обработчик событий, то удобнее хранить все данные формы в одном стейте.
```
	const [formState, setFormState] = useState(defaultArticleState);

	const handleChange = (type: keyof ArticleStateType, value: OptionType) => {
		setFormState({
			...formState,
			[type]: value
		})
	};
	```

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,14 +1,67 @@
+				ref={containerRef}
+				className={clsx(styles.container, props.open && styles.container_open)}
+				onClick={(ev) => {
+					ev.stopPropagation();
```
**Можно лучше:** 
Для параметра являющегося событием общеупотребимым сокращением названия является `evt` или `e` лучше придерживаться этого имени.

olgalatkina reviewed (changes requested):
Добрый день!

**У вас получилась хорошая работа:**
1. Полностью реализован весь функционал.
2. Проект соответствует макету на сколько это возможно.
3. Для условного добавления css-классов спользуется библиотека `clsx`.
4. Логика закрытия сайдбара по клику вне сайдбара вынесена в кастомный хук.
5. Обработчик сохранения данных вызывается при отправке формы.

**Но есть несколько моментов, которые надо поправить:**
- [ ] Компонент страницы не должен знать про состояние формы. Состояние формы надо перенести в компонент `ArticleParamsForm`. Туда же нужно перенести все обработчики.
- [ ] Не нужно создавать компонент `InputArticleParams` и переменную `defaultStyle`
- [ ] В проекте как настройки по умолчанию нужно использовать значения из объекта `defaultArticleState`
- [ ] Исправить кастомный хук, чтобы обработчик закрытия по клику вне сайдбара  навешивался при его открытии и удалялся при закрытии.
 
**Правки можно лучше остаются на ваше усмотрение:**
- [ ] Стоит вынести компонент `App` в отдельный файл в папке `components/app/App.tsx`.
- [ ] Для одиночных селекторов не стоит использовать библиотеку `clsx `(см. компоненты `Article` и `App`).
- [ ] Попробуйте обойтись без лишней обёртки `toggleOpen` над функцией-сеттером
- [ ] Используйте деструктуризацию пропсов сразу в параметрах компонента
 
Успешного рефакторинга!

aleoreh committed:
Переносит "gap" в основной класс ".form"

- В этой проектной работе нет необходимости писать
дополнительный css-класс. Свойство gap просто добавляем в класс .form.

aleoreh committed:
Переключает состояние по умолчанию

Настройки статьи по умолчанию описаны в объекте defaultArticleState
в файле constants/articleProps.ts стартеркита.
В проекте нужно использовать значения из этого объекта
как настройки по умолчанию, как для самой статьи,
так и как начальные значения полей форм.

aleoreh committed:
Переключает InputArticleParams на ArticleStateType

В проекте как настройки по умолчанию нужно использовать значения
из объекта defaultArticleState.
Этот тип (ArticleStyle) стоит удалить как неиспользуемый.

aleoreh committed:
Выносит App в отдельный модуль

-
Обычно корневой index.tsx держат максимально компактным.
Стоит вынести компонент App в отдельный файл в папке
components/app/App.tsx.

aleoreh committed:
Перемещает InputArticleParams в ArticleParamsForm

-
Поля формы это часть компонента ArticleParamsForm,
отвечающего за настройки отображения статьи и не предусмотрено,
что в форму будет передаваться разный набор полей ввода.
В этой проектной работе всё наполнение формы прописываем в компоненте
ArticleParamsForm. Просто собираем её из готовых компонентов.
Компонент InputArticleParams не понадобится.

aleoreh committed:
Переносит видимость формы в ArticleParamsForm

-
Обычно переменные, которые сохраняют булево значение
начинаются с глагола is или has, например, isOpen
Это состояние тоже нужно перенести в компонент ArticleParamsForm
вместе обработчиками отправки и сброса формы.

aleoreh committed:
Улучшает типизацию события

-
Можно лучше:
React.FormEvent<HTMLFormElement>

aleoreh committed:
Убирает ненужную передачу ссылки на кнопку

-
Использование forwardRef избыточно.
Достаточно передать флаг открытости/закрытости сайдбара
и функцию меняющую этот флаг.

aleoreh committed:
Деструктуризирует параметры компонентов

-
Можно использовать деструктуризацию пропсов
сразу в параметрах компонента, это распространенная практика.

aleoreh committed:
Убирает ненужную обертку над обработчиком

-
Нет необходимости создавать дополнительные функции обработчики
если они сами по себе ничего не делают,
можно использовать просто функции сеттеры возвращаемые useState.

aleoreh committed:
Улучшает типизацию ссылки

-
Реф инициализируется значением null, поэтому типизация будет выглядеть
так:
const containerRef = useRef<HTMLElement | null>(null);

aleoreh committed:
Заменяет useOutsideClick на useModalClose

-
Используйте событие mousedown
и тогда не придётся использовать evt.stopPropagation()
у компонента ArrowButton.

aleoreh committed:
Улучшены имена параметров обработчков событий

-
Для параметра являющегося событием общеупотребимым сокращением названия
является evt или e лучше придерживаться этого имени.

aleoreh committed:
Удаляет ненужную обработку события

aleoreh committed:
Немного чистит код и добавляет комментарии

aleoreh committed:
Исправляет ошибку "панель не закрывается"

-
Возникала при быстром клике на кнопку закрытия
панели тогда, когда она находится в открытом состоянии

aleoreh committed:
Revert "Исправляет ошибку "панель не закрывается""

This reverts commit a7d0c83b27a7ff1d77cd6cbd95df7759426ee4e0.

olgalatkina commented on src/components/app/App.tsx:
```diff
@@ -0,0 +1,46 @@
+		useState<ArticleStateType>(defaultArticleState);
+
+	// при утверждении формы ввода параметров устанавливаем их в стили статьи
+	const submitArticleParamsForm = (articleState: ArticleStateType) => {
```
**Надо исправить:**
Не исправлено замечание с прошлого ревью: обработчики событий формы `submitArticleParamsForm` и `onArticleParamsFormReset` надо перенести в компонент `ArticleParamsForm`. Страница ничего не должна знать про форму и её события. В вашем решении в компонент `ArticleParamsForm` достаточно пробросить функцию-сеттер, возвращаемую хуком `useState`. 

В React ссылка на функцию-сеттер, возвращаемую хуком `useState`, обычно является стабильной. Это означает, что она будет оставаться той же самой функцией при каждом рендере компонента.

olgalatkina commented on src/hooks/useModalClose.ts:
```diff
@@ -0,0 +1,43 @@
+	useEffect(() => {
+		if (!isOpened) return;
+
+		const normalizedSelfRefs = Array.isArray(selfRefs) ? selfRefs : [selfRefs];
```
**Надо исправить:**
Вряд ли здесь может оказаться массив. Стоит исправить этот момент.

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,14 +1,116 @@
+				onClick={() => setIsOpened(!isOpened)}
+			/>
+			<aside
+				ref={containerRef}
```
**Можно лучше:**
Если реф перенести на внешний `div`, то баг с кнопкой уйдёт. `div` будет на месте `React.Fragment`. Не забудьте протипизировать хук `useRef`.

olgalatkina reviewed (changes requested):
Отличная работа!

**Осталось исправить совсем немного:**
- [ ] Перенести обработчики событий формы в `ArticleParamsForm`
- [ ] Поправить кастомный хук

**Можно лучше:**
Стоит перенести реф на внешний `div`.

Успешного рефакторинга!

aleoreh committed:
Исправляет неправильное поведение панели

-
При "быстром" клике на кнопку закрытия панель начинает скрываться,
а затем на полпути открывается обратно

aleoreh committed:
Изменяет useModalClose: принимает один Ref

-
Вряд ли здесь может оказаться массив. Стоит исправить этот момент.

aleoreh committed:
Переносит применение стиля в ArticleParamsForm

-
обработчики событий формы submitArticleParamsForm
и onArticleParamsFormReset надо перенести в компонент ArticleParamsForm

aleoreh committed:
Небольшое изменение имени

olgalatkina reviewed (approved):
Добрый день!

**Отличная работа! проект принимается**🎉

Успехов в дальнейшем обучении!
